RoR Capstone project

Wahaj-Ali opened the pull request:
Hi 👋 

## HIGHLIGHTS of the project
This web app has the following features:
#### Interactions
- Sign up and login pages.
- Home page (categories page).
- Transactions page.
- For each category, the user can see the category name, icon, and total amount of all the transactions that belong to the category.
- Two forms to add new categories or new transactions.

There is a button "add transaction" at the bottom that brings the user to the page to create a new transaction, and upon saving it adds a new transaction to the transactions list and redirects the user to the home page.
When the user clicks on the "Back" button (<), the user navigates to the home page.

Similarly, the "Add Category" button creates a new category, and upon saving it adds a new category to the categories list and redirects the user to the home page.
When the user clicks on the "Back" button (<), the user navigates to the home page.

- Unit & Integraton test

Add unit and integration tests for all the most important components using RSpec.

- Deployment

Use Render to deploy the App. [Link](https://budget-app-eydq.onrender.com/)

#### Technical requirements
- Use Postgres as your database.
- Use devise for authentication.
- Add validations so that no one with bad intentions can compromise your app.
- Project is deployed and accessible online.
- The database scheme matches the given ERD diagram.

[Project presentation Video](https://www.loom.com/share/c9731690e7044d419e01962d1f4b598a) (UPDATED)

Wahaj-Ali committed:
Set up new rails app using postgres.

Wahaj-Ali committed:
Set-up GitHub Actions & Set-up linters for local enviornment.

In this commit
- Remove linter errors.

Wahaj-Ali committed:
Merge pull request #1 from Wahaj-Ali/Rails-setup

Rails setup

Wahaj-Ali committed:
Add devise

In this commit
- Use devise for user authentication.
- Add a navbar with log in and log out buttons.
- Add a route to the home page that shows the name of the App.

Wahaj-Ali committed:
Remove linter errors.

Wahaj-Ali committed:
Merge pull request #2 from Wahaj-Ali/User-authentication

User authentication

Wahaj-Ali committed:
Add a role column in the users table.

Wahaj-Ali committed:
Add migration to create groups table.

Wahaj-Ali committed:
Add user reference to groups using rails migration.

Wahaj-Ali committed:
Add money tracks table with reference to the users.

Wahaj-Ali committed:
Create money groups table.

Wahaj-Ali committed:
Delete 20230425143222_add_user_to_money_groups.rb

Wahaj-Ali committed:
Add association between Money group, money tracks and group.

Wahaj-Ali committed:
Update schema.rb

Wahaj-Ali committed:
Edit navbar

Wahaj-Ali committed:
Merge pull request #3 from Wahaj-Ali/migrations

Migrations

Wahaj-Ali committed:
Install Cancancan gem for authorization of the user.

Wahaj-Ali committed:
Create groups_controller.rb

Wahaj-Ali committed:
Create money_tracks_controller.rb

Wahaj-Ali committed:
Update application.css

Wahaj-Ali committed:
Update user.rb

Wahaj-Ali committed:
Create money_track.rb

Wahaj-Ali committed:
Create money_group.rb

Wahaj-Ali committed:
Create group.rb

Wahaj-Ali committed:
Add styles to the sign up, sign in and edit profile forms.

Wahaj-Ali committed:
Add views for groups pages.

Wahaj-Ali committed:
Update application.html.erb

Wahaj-Ali committed:
Add views for money_tracks

Wahaj-Ali committed:
Update _navbar.html.erb

Add styling to the navbar.

Wahaj-Ali committed:
Update routes.rb

Wahaj-Ali committed:
Remove linter errors.

Wahaj-Ali committed:
Merge pull request #4 from Wahaj-Ali/auhtorization

Auhtorization

Wahaj-Ali committed:
Update application.css

Wahaj-Ali committed:
Edit sign up form for a new user.

In this commit
- Add a navbar foe the sign-up form.

Wahaj-Ali committed:
Edit sign-in form.

In this commit
- Add a navbar for the sign-in form.

Wahaj-Ali committed:
Update the views (index) for the categories.

Wahaj-Ali committed:
Update application_controller.rb

In this commit
- remove authenticate_user so that the home index can be shown properly.

Wahaj-Ali committed:
Update groups_controller.rb

Add destroy method for the groups controller.

Wahaj-Ali committed:
Update money_tracks_controller.rb

Add destroy method for the money tracks controller.

Wahaj-Ali committed:
Update users_controller.rb

Wahaj-Ali committed:
Update views for groups.

Wahaj-Ali committed:
Create _errornotice.html.erb

Add the notice to the layouts to render it wherever needed.

Wahaj-Ali committed:
Update application.html.erb

Remove notices.

Wahaj-Ali committed:
Update the view of money transactions page.

Wahaj-Ali committed:
Create index.html.erb

Add views for the home-page of the App.

Wahaj-Ali committed:
Update routes.rb

Wahaj-Ali committed:
Remove styleint errors.

Wahaj-Ali committed:
Merge pull request #5 from Wahaj-Ali/styles

Styles & Views

Wahaj-Ali committed:
Add style to the log out icon.

Wahaj-Ali committed:
Add image icons for different categories & add/edit styles.

In this commit
- Use bootstrap classes to style the app.
- Use images for the icons of different categories.

Wahaj-Ali committed:
Edit styles

In this commit
- Edit styles for sign-up form.
- Edit styles for sign-in form.
- Edit styles for category index page.

Wahaj-Ali committed:
Merge pull request #6 from Wahaj-Ali/final-edits

Final edits

Wahaj-Ali committed:
Install RSpec in the App.

Wahaj-Ali committed:
Write integration test for index page of categories.

Wahaj-Ali committed:
Add integration test for show page of categories.

Wahaj-Ali committed:
Add integration test for adding a new category page.

Wahaj-Ali committed:
Add integration test for adding a new transaction form/page.

Wahaj-Ali committed:
Update new.html.erb

Modify the grammatical error in the heading.

Wahaj-Ali committed:
Create group_spec.rb

Add test for the category model.

Wahaj-Ali committed:
Create money_track_spec.rb

Add test for transaction model.

Wahaj-Ali committed:
Create user_spec.rb

Add a test for the user model.

Wahaj-Ali committed:
Update user model to assign role if not present.

Wahaj-Ali committed:
Create groups_spec.rb

Add a test for the groups controller.

Wahaj-Ali committed:
Create money_tracks_spec.rb

Add test for the transactions(money_tracks) controller.

Wahaj-Ali committed:
Modify the styles for a better UI.

- Use bootstrap classes to add styles.
- Use in-line CSS as well.

Wahaj-Ali committed:
Update ability.rb

If the user is the creator of the category, he can delete that category.

Wahaj-Ali committed:
Remove linter errors.

Wahaj-Ali committed:
Update application.css

Add media query for the desktop version.

Wahaj-Ali committed:
Update README.md

Add a descriptive Readme file.

Wahaj-Ali committed:
Merge pull request #7 from Wahaj-Ali/tests

Unit & Integration Tests.

Wahaj-Ali committed:
Update database.yml

Update database according to the Render docs to deploy Rails App.

Wahaj-Ali committed:
Update puma.rb

Wahaj-Ali committed:
Update production.rb

Wahaj-Ali committed:
Create render-build.sh

Wahaj-Ali committed:
Create render.yaml

Wahaj-Ali committed:
Update credentials.yml.enc

Wahaj-Ali committed:
Update render.yaml

Wahaj-Ali committed:
Update Gemfile.lock

Wahaj-Ali committed:
Update README.md

Add live demo link.

Wahaj-Ali committed:
Update README.md

Add a link to a video presentation of the project.

Wahaj-Ali committed:
Update README.md

Add design acknowledgment.

Wahaj-Ali committed:
Merge pull request #8 from Wahaj-Ali/deployment

Deployment

codepantha commented on app/views/money_tracks/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+
+  <nav class="header">
+  <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+    <li>
+    <a href="/"><i class="fa fa-arrow-left"></i></a>
+    </li>
```
- [x] From the "Add Transaction" page, when a user clicks on the "Back" button (<), the user should navigate to the transactions page for that category. At the moment, the "Back" button (<) button takes us to the homepage. Please make this change 🙏

Wahaj-Ali replied:
@codepantha thank you for pointing out this issue, surely making this change will make my App better.

Wahaj-Ali replied:
@codepantha I found the font used in the design. I have updated my App as well.

codepantha commented on app/views/groups/index.html.erb:
```diff
@@ -0,0 +1,45 @@
+  <div class="p-2 m-3  category-div">
+   <div class="category-info">
+    <p> <%=image_tag "#{group.icon}" ,class:"p-2 m-1 rounded img-thumbnail"%></p>
+    <ul class="p-2 m-1">
+      <li> <p class="d-flex cat_name"><%= link_to group.name, group_path(group), class:"p-2 m-2 text-muted h2" %></p></li>
+      <li> <p class="p-2 m-2 text-muted cat_date"> <%= group.created_at.strftime("%d-%b-%Y")%></p></li>
+    </ul>
+    </div>
+    <ul class="p-2 m-1">
+    <li><p class="p-2 m-1 h3"> $<%= group.money_tracks.sum(:amount)%></p></li>
+    <li>
```
- [x] Please follow the design guidelines in terms of `font-family` and `font-weight` here and across your application.
- [x] I also noticed that the category's name is cut off even though there's more than enough room to contain it. Please try to fix it in order to improve your UI 👍 
### Design Guideline
![image](https://user-images.githubusercontent.com/46839548/234674616-44acb068-f31f-46b7-99ee-9448308e13b8.png)

### Your Design
![image](https://user-images.githubusercontent.com/46839548/234676055-0ee64f4a-ab20-4128-b14d-613c860df216.png)

Wahaj-Ali replied:
Hi @codepantha, I tried to find the font on google fonts, but it is not available. So I won't be able to use the exact font family in my App.

Wahaj-Ali replied:
@codepantha I found the font used in the design. I have updated my App as well.

codepantha reviewed (changes requested):
# STATUS: CHANGES REQUIRED ♻️ 

## Hi @Wahaj-Ali 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media3.giphy.com/media/rZ10JtoPRv1Ly/giphy.gif"/>

## Highlights 🎯 
✔️ Descriptive PR
✔️ Good commit history
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

- [ ] According to the requirements, the video should contain a highlight about a piece of the code that you find interesting or are proud of making. Kindly re-do the video and incorporate this suggestion 👍 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Wahaj-Ali committed:
Update new.html.erb

- As required, now the back button from the 'Add Transaction' page takes the user back to the category details(transactions) page.

Wahaj-Ali committed:
As required, edit styles to improve the UI by adjusting the font-sizes and making changes to adjust the 'categories' names ' to be shown completely.

Wahaj-Ali committed:
Update application.css

Wahaj-Ali commented:
Thank you @codepantha for your valuable suggestions, I am sure these suggestions will make it the best version of my App.

Wahaj-Ali committed:
Update styles.

In this commit
- Update font-family to match what is used in the design.
- Update the dimensions of different parts of the App to match the design.

Wahaj-Ali committed:
Update show.html.erb

Removed margin bottom form the navbar.

Wahaj-Ali committed:
Update money_tracks_controller.rb

Update the transaction controller, now when the user creates a new transaction and saves it, it redirects to the specific category page in which the transaction was created.

Wahaj-Ali committed:
Update README.md

Update project presentation video link.

sumairq reviewed (approved):
### Hi @Wahaj-Ali , 

<img src="https://media3.giphy.com/media/BPJmthQ3YRwD6QqcVD/giphy.gif"/>

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

## Highlights
✔️ Good commit history
✔️Linters are passing
✔️Very professional README
✔️Correct Git flow
✔️Descriptive PR


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
